File: app/src/main/java/net/squanchy/venue/VenueInfoService.java
package net.squanchy.venue;

import net.squanchy.service.firebase.FirebaseAuthService;
import net.squanchy.service.repository.VenueRepository;
import net.squanchy.venue.domain.view.Venue;

import io.reactivex.Observable;

class VenueInfoService {

    private final VenueRepository venueRepository;
    private final FirebaseAuthService authService;

    VenueInfoService(VenueRepository venueRepository, FirebaseAuthService authService) {
        this.venueRepository = venueRepository;
        this.authService = authService;
    }

    public Observable<Venue> venue() {
        return authService.ifUserSignedInThenObservableFrom(userId -> venueRepository.venue());
    }
}

File: app/src/main/java/net/squanchy/venue/VenueInfoService.kt
package net.squanchy.venue

import io.reactivex.Observable
import net.squanchy.service.firebase.FirestoreDbService
import net.squanchy.service.firebase.model.conferenceinfo.FirestoreVenue
import net.squanchy.service.repository.AuthService
import net.squanchy.venue.domain.view.Venue
import net.squanchy.venue.domain.view.toVenue

internal class VenueInfoService(private val dbService: FirestoreDbService, private val authService: AuthService) {

    fun venue(): Observable<Venue> {
        return authService.ifUserSignedInThenObservableFrom {
            dbService.venueInfo()
                .map(FirestoreVenue::toVenue)
        }
    }
}
